src/NuGet.Jobs.Catalog2Registration/Hives/HiveMerger.cs

```diff
@@ -0,0 +1,311 @@
+            // See the wonderful Wikipedia page: https://en.wikipedia.org/wiki/Merge_algorithm
+
+            var catalogIndex = 0;
+            var pageIndex = 0;
+
+            var sortedCatalog = context.SortedCatalog;
+            var pageInfos = context.IndexInfo.Items;
+
+            // Keep track of the position in the current page.
+            var itemIndex = 0;
+
+            // This is the main "merge" step where the two input lists are interleaved.
+            while (catalogIndex < sortedCatalog.Count && pageIndex < pageInfos.Count)
+            {
+                var catalog = sortedCatalog[catalogIndex];
+                var pageInfo = pageInfos[pageIndex];
+                var nextLower = pageIndex < pageInfos.Count - 1 ? pageInfos[pageIndex + 1].Lower : null;
+                var hasGap = pageInfo.Count < MaxLeavesPerPage && nextLower != null && catalog.PackageIdentity.Version < nextLower;
+
+                if (catalog.PackageIdentity.Version <= pageInfo.Upper || hasGap)
+                {
+                    itemIndex = await MergeEntryIntoPageAsync(context, catalog, pageIndex, itemIndex);
+                    catalogIndex++;
+                }
+                else
+                {
+                    // Before considering the current page "complete" ensure the page size is correct.
+                    await EnsureValidPageSizeAtAsync(context, pageIndex);
+
+                    if (catalog.PackageIdentity.Version > pageInfo.Upper)
+                    {
+                        // The current catalog entry is greater than the current page's upper bound. This means we're done
+                        // with the current page. Move to the next page and reset the item index.
+                        pageIndex++;
+                        itemIndex = 0;
+                    }
+                }
+            }
+
+            // Now that one of the two input lists is drained, handle the other (undrained) list. 
+
+            // Process the rest of the catalog leaves, if any.
+            if (catalogIndex < sortedCatalog.Count)
+            {
+                // Make sure there is at least one non-full page so that remaining catalog leaves can be pushed there.
+                if (pageInfos.Count == 0
+                    || pageInfos.Last().Count == MaxLeavesPerPage)
+                {
+                    context.IndexInfo.Insert(pageIndex, PageInfo.New());
+                }
+                else
+                {
+                    pageIndex = pageInfos.Count - 1;
+                }
+
+                // Push the remaining catalog leaves into the last page.
+                while (catalogIndex < sortedCatalog.Count)
+                {
+                    var catalog = sortedCatalog[catalogIndex];
+                    itemIndex = await MergeEntryIntoPageAsync(context, catalog, pageIndex, itemIndex);
+                    catalogIndex++;
+                }
+                RemovePageAtIfEmpty(context, pageIndex);
```

Comment:
Nit: new line between 135 and 136? Feel free to ignore if this is intentional

Reply:
Not intentional, good catch.